src/stream/src/executor/source.rs

```diff
@@ -350,30 +390,40 @@ impl<S: StateStore> SourceExecutor<S> {
-                                            );
-                                            let reader = self
-                                                .build_stream_source_reader(Some(
-                                                    target_state.clone(),
-                                                ))
-                                                .await
-                                                .map_err(StreamExecutorError::source_error)?;
-                                            abort_tx.send(()).unwrap();
-                                            *stream_reader.lock().await = Some(reader);
-                                            self.stream_source_splits = target_state;
+                            if let Some(mutation) = barrier.mutation.as_deref() {
+                                match mutation {
+                                    Mutation::SourceChangeSplit(mapping) => {
+                                        if let Some(target_splits) =
+                                            mapping.get(&self.actor_id).cloned()
+                                        {
+                                            if let Some(target_state) = self.get_diff(target_splits)
+                                            {
+                                                log::info!(
+                                                    "actor {:?} apply source split change to {:?}",
+                                                    self.actor_id,
+                                                    target_state
+                                                );
+                                                let reader = self
+                                                    .build_stream_source_reader(Some(
+                                                        target_state.clone(),
+                                                    ))
+                                                    .await
+                                                    .map_err(StreamExecutorError::source_error)?;
+                                                abort_tx
+                                                    .send(AbortMessage::SourceChange(()))
+                                                    .unwrap();
+                                                *stream_reader.lock().await = Some(reader);
+                                                self.stream_source_splits = target_state;
+                                            }
                                         }
                                     }
+                                    Mutation::Pause(_) => {
+                                        abort_tx.send(AbortMessage::Pause(())).unwrap();
+                                    }
+                                    Mutation::Resume(_) => {
+                                        abort_tx.send(AbortMessage::Resume(())).unwrap();
+                                    }
```

Comment:
Do we really need to make the source reader aware of `Pause` and `Resume`? IMHO, if we do not poll the reader here, then we're already paused.

Reply:
yes, but polling source reader happens in a separate thread, we should tell it when to pause and when to resume